chain/client/src/client.rs

```diff
@@ -485,6 +471,90 @@ impl Client {
         Ok(Some((encoded_chunk, merkle_paths, outgoing_receipts)))
     }
 
+    /// Pulls transactions from the transaction pool for a given shard up to the gas_limit.
+    /// Filters transactions one by one by persisting the state across valid validations.
+    fn get_filtered_transactions(
```

Comment:
As discussed, let's move `get_filtered_transactions` into the `RuntimeAdapter`. A method of `RuntimeAdapter` can receive the draining iterator and a closure `check_relevant_to_chain` that checks whether a given transaction is still on the same chain.